geomstats/learning/frechet_mean.py

```diff
@@ -158,27 +159,74 @@ def _default_gradient_descent(points, metric, weights,
     return mean
 
 
-def _ball_gradient_descent(points, metric, weights, max_iter,
+def _ball_gradient_descent(points, metric, weights=None, max_iter=32,
                            lr=1e-3, tau=5e-3):
     if len(points) == 1:
         return points
 
-    iteration = 0
-    convergence = math.inf
-    barycenter = points.mean(0, keepdims=True) * 0
+    if weights is None:
 
-    while convergence > tau and max_iter > iteration:
+        iteration = 0
+        convergence = math.inf
+        barycenter = points.mean(0, keepdims=True) * 0
+        while convergence > tau and max_iter > iteration:
 
-        iteration += 1
+            iteration += 1
+
+            grad_tangent = 2 * metric.log(points, barycenter)
+
+            cc_barycenter = metric.exp(
+                lr * grad_tangent.sum(0, keepdims=True), barycenter)
+
+            convergence = metric.dist(cc_barycenter, barycenter).max().item()
+
+            barycenter = cc_barycenter
+    else:
+        lr = 5e-2
+        tau = 5e-3
+
+        weights = gs.expand_dims(weights, -1)
+        weights = gs.repeat(weights, points.shape[-1], axis=2)
+
+        barycenter = (points * weights).sum(0, keepdims=True) / weights.sum(0)
+        iteration = 0
+        convergence = math.inf
+        points_gs = gs.squeeze(points)
+
+        barycenter_gs = gs.squeeze(barycenter)
+
+        while convergence > tau and max_iter > iteration:
+
+            iteration += 1
+
+            grad_tangent = gs.zeros((len(points),
+                                     len(barycenter_gs), len(points[0][0])))
+
+            # TODO Hadi/Thomas: change exp/log to handle arrays
+
+            for j, _ in enumerate(points):
```

Comment:
for j, point in enumerate(points_gs)?

Reply:
DeepSource was reporting an error if I use 'for j, points' instead of 'for j,_' for unused variable point

Reply:
yes, but then you can use `point` instead of `point_gs[j]` and then `point` is used.

Reply:
Indeed! sorry. These loops have been removed with the vectorization but I've applied it for enumerates in plot_expectation_maximisation_manifolds.py :)